### PR TITLE
#1067 - Datasets

### PR DESCRIPTION
**Pull request by JohnPTobe**
##### Checklist

- [x] `manage.py test` passes
- [x] tests are included
- [x] migrations are included
- [x] documentation is changed or added

### Affected app(s)
- data
- dataset

### Description of change
Adding dataset model and associated api to Scale.  I wasn't sure what fields to include in various endpoints so if something seems like it would be useful and is missing let me know.  Also, I'm considering moving everything under data if it makes sense.

### CODE REVIEW

**Comment by cshamis** [comment]
I worry about storing definitions of datasets... the pattern is to just get a static storage model that can hold groups of scale files.  

Not "dynamic playlists" although... that's an AWESOME IDEA for S8!

File: scale/dataset/models.py [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+
+logger = logging.getLogger(__name__)
+
+DataSetValidation = namedtuple('DataSetValidation', ['is_valid', 'errors', 'warnings'])
+# DataSetKey = namedtuple('DataSetKey', ['name', 'version'])
+
+class DataSetManager(models.Manager):
+    """Provides additional methods for handling datasets"""
+
+    @transaction.atomic
```
[the remark this answers is not in the document]
Comment (JohnPTobe) [reply scale/dataset/models.py]: I'll remove the decoration as it's not needed.  Just left over from a copy/paste. We're not updating any objects that exist outside this method.

File: scale/dataset/models.py [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+
+        if not data:
+            raise InvalidDataSetMember('INVALID_DATASET_MEMBER', 'No data provided')
+
+        try:
+            dataset.get_definition().validate(data)
+        except InvalidData as ex:
+            raise InvalidDataSetMember('INVALID_DATASET_MEMBER', 'Data does not match dataset parameters: %s' % ex)
+
+        dataset_member = DataSetMember()
```
[the remark this answers is not in the document]
Comment (JohnPTobe) [reply scale/dataset/models.py]: Yes, this needs to be atomic as we're creating the datasetfiles and member object at the same time. Good catch!

File: scale/dataset/models.py [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+        file_ids = []
+        for i in data.values.keys():
+            v = data.values[i]
+            if type(v) is FileValue:
+                file_ids.extend(v.file_ids)
+                for id in v.file_ids:
+                    exists = DataSetFile.objects.filter(dataset=dataset, scale_file_id=id).count()
+                    if exists:
+                        continue
+                    file = DataSetFile()
```
[the remark this answers is not in the document]
Comment (JohnPTobe) [reply scale/dataset/models.py]: Same as above

File: scale/dataset/models.py [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+                    file.dataset = dataset
+                    file.scale_file = ScaleFile.objects.get(pk=id)
+                    file.parameter_name = i
+                    file.save()
+        return file_ids
+
+    def get_file_ids(self, dataset_ids, parameter_name=None):
```
[the remark this answers is not in the document]
Comment (JohnPTobe) [reply scale/dataset/models.py]: Maybe. It'd be simple enough to change.

**Comment by JohnPTobe** [comment]
@emimaesmith I think I addressed everything.
@cshamis The definition allows us to create groups of files for multiple inputs.  Each member of a dataset needs to fulfill the interface defined by the dataset and thus you can be assured when passing a dataset to a recipe that all the members will have the required inputs.

**Pull request merged**